src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,159 @@
+import cn from 'classnames';
+import {
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import { Todo } from '../../types/Todo';
+import { AppContext } from '../../AppContext';
+
+interface Props {
+  todo: Todo;
+  removeTodo: (todo: Todo) => Promise<void>;
+  isLoading?: boolean;
+  updateTodo: (todo: Todo) => Promise<void>;
+}
+
+export const TodoItem = ({
+  todo,
+  removeTodo,
+  isLoading,
+  updateTodo,
+}: Props) => {
+  const { state } = useContext(AppContext);
+
+  const [loaderStatus, setLoaderStatus] = useState(isLoading);
+  const [isEditing, setIsEditing] = useState(false);
+  const [editingTitle, setEditingTitle] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+  const formRef = useRef<HTMLFormElement>(null);
+
+  useEffect(() => {
+    inputRef.current?.focus();
+  }, [isEditing]);
+
+  const handleCheckboxClick = () => {
+    setLoaderStatus(true);
+    updateTodo({
+      ...todo,
+      completed: !todo.completed,
+    })
+      .finally(() => setLoaderStatus(false));
+  };
+
+  const handleDeleteTodo = () => {
+    setLoaderStatus(true);
+    removeTodo(todo)
+      .then(() => {
+        setIsEditing(false);
+        setLoaderStatus(false);
+      });
+  };
+
+  const handleTitleClick = (event: React.MouseEvent<HTMLSpanElement>) => {
+    if (event.detail >= 2) {
+      setIsEditing(true);
+    }
+  };
+
+  const handleInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setIsEditing(false);
+      setEditingTitle(todo.title);
+    }
+  };
+
+  const handleInputSubmit = (event?: React.FormEvent) => {
+    event?.preventDefault();
+
+    if (editingTitle === todo.title) {
+      setIsEditing(false);
+
+      return;
+    }
+
+    const editedTodo: Todo = {
+      ...todo,
+      title: editingTitle.trim(),
+    };
+
+    if (editedTodo.title === '') {
+      handleDeleteTodo();
+
+      return;
+    }
+
+    setLoaderStatus(true);
+    updateTodo(editedTodo)
+      .finally(() => {
+        setIsEditing(false);
+        setLoaderStatus(false);
+      });
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn('todo', { completed: todo.completed })}
+      key={todo.id}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={handleCheckboxClick}
+        />
+      </label>
+
+      {isEditing ? (
+        <form onSubmit={handleInputSubmit} ref={formRef}>
```

Comment:
I think it would be great if after submitting with update error this form didn't close and stayed focused
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/147749101/45d4179f-12fc-423b-bb44-3f98e2bec6d2)

like here: 
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/147749101/694401b0-894d-41a7-a8a7-0fc008e7c35d)